prober/derp.go

```diff
@@ -577,6 +692,88 @@ func runDerpProbeNodePair(ctx context.Context, from, to *tailcfg.DERPNode, fromc
 	return nil
 }
 
+func runDerpProbeNodePairContinuously(ctx context.Context, from, to *tailcfg.DERPNode, fromc, toc *derphttp.Client, pktSize int64, bytesTransferred, transferTime *expvar.Float, qdh *queuingDelayHistogram) error {
+	// To avoid derper dropping enqueued packets, limit the number of packets in flight.
+	// The value here is slightly smaller than perClientSendQueueDepth in derp_server.go
+	sentTimes := make(chan time.Time, 30)
+
+	// Send the packets.
+	sendc := make(chan error, 1)
+	pkt := make([]byte, pktSize) // sized slightly smaller than MTU to avoid fragmentation
+	crand.Read(pkt)
+	go func() {
+		for {
+			now := time.Now()
+			sentTimes <- now
+			if err := fromc.Send(toc.SelfPublicKey(), pkt); err != nil {
+				sendc <- fmt.Errorf("sending packet %w", err)
+				return
+			}
+		}
+	}()
+
+	// Increment transfer time every 1 second.
+	ticker := time.NewTicker(1 * time.Second)
+	go func() {
+		for {
+			select {
+			case <-ticker.C:
+				transferTime.Add(1)
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+
+	// Receive the packets.
+	recvc := make(chan error, 1)
+	go func() {
+		defer close(recvc) // to break out of 'select' below.
+		for {
+			m, err := toc.Recv()
+			if err != nil {
+				recvc <- err
+				return
+			}
+			switch v := m.(type) {
+			case derp.ReceivedPacket:
+				now := time.Now()
+				sentTime := <-sentTimes
+				if v.Source != fromc.SelfPublicKey() {
+					recvc <- fmt.Errorf("got data packet from unexpected source, %v", v.Source)
+					return
+				}
+				if !bytes.Equal(v.Data, pkt) {
+					recvc <- fmt.Errorf("unexpected data packet")
+					return
+				}
+				qdh.add(now.Sub(sentTime))
```

Comment:
Did I misread [this](https://github.com/tailscale/tailscale/blob/main/prober/derp.go#L547)? It looked to me like the existing probe expects packets to arrive in the order they were sent.

Reply:
Same assumption is incorrect there. One of the reasons it mostly works is the fixed (30 packets) flow control being applied.

Reply:
Haha, okay. I've

1. Added a comment to the existing code that warns about this behavior

2. For the continuous probe, I'm marshaling the monotonic time into the packet headers. I had to do some reflection for this because the monotonic time data isn't exposed by the public time.Time methods. AFAICT, it's safe to reuse the single byte slice for the packet as each call to `Send()` ends up copying it anyway.